### PR TITLE
Enable user-defined injection for pod annotations

### LINKED ISSUE #57: Support customized injection
Sometimes, user wants to inject customized content in pod manifest on the fly. For example, adding an additional network by injecting net-attach-defs in pod annotation.

This can be achieved by defining a configMap which contains key:value in its data entry, key is a user defined k8s label, value is the content to be injected in pod manifest. NRI inspects the pod label and inject the corresponding data value in pod manifest if the pre-defined label is found. For example:

```
apiVersion: v1
data:
  network-resource-injector-pod-annotation: '{"op": "add", "path": "/metadata/annotations", "value": {"k8s.v1.cni.cncf.io/networks": "sriov-net-attach-def"} }'
kind: ConfigMap
metadata:
  name: network-resource-injector-customized-injection-config
  namespace: network-resource-injector-namespace
```
In above example, `network-resource-injector-pod-annotation` is the k8s label defined by user,  followed by the value (json string) defining customized content to be injected in pod spec.

NRI would listen to the configMap change and record the configmap data entry in its internal data structure. 
During pod creation, NRI receives the pod manifest and inspects the pod labels, if any of pod label is found to match with the key of recorded data entry, NRI would inject the corresponding data entry value to pod manifest.
If configMap is not provided or data is empty in configMap, NRI would skip the customized injection.

### PR DESCRIPTION
**Pull request by zshi-redhat**
User-defined injections allows user to define additinal injections in
Kubernetes ConfigMap and request additional injection for individual
pod based on pod label. Currently user-defined injection only support
injecting pod annotations.

Fixes #57 
Signed-off-by: Zenghui Shi <zshi@redhat.com>

### CODE REVIEW

File: pkg/webhook/webhook.go
Comment (pperiyasamy) [inline pkg/webhook/webhook.go:587]: Nit: can't we just log annotateKey instead of netType ? this way networkAnnotationKey can be passed from calling side (instead of netType) which avoids above switch case.

File: cmd/webhook/main.go
Comment (pperiyasamy) [inline cmd/webhook/main.go:35]: would this be only one global config map for all the pods  ?  should we update same existing config map if there is a need to add/update/delete data field ?

File: pkg/webhook/webhook.go
Comment (zshi-redhat) [reply pkg/webhook/webhook.go]: Yes, good catch! updated.

File: cmd/webhook/main.go
Comment (zshi-redhat) [reply cmd/webhook/main.go:35]: Yes, this will be the only global configmap for all pods. A pod can decide which data field shall be applied by using different labels.   This single configmap shall be updated when customized injection will be added/updated/deleted.

File: pkg/webhook/webhook.go
Comment (pperiyasamy) [inline pkg/webhook/webhook.go:899]: shouldn't we also remove stale entries from customizedInjections Map (to handle delete cases in config map) ?

Comment (zshi-redhat) [reply pkg/webhook/webhook.go:899]: Yes, we should.
Done

**Comment by pperiyasamy** [comment]
/lgtm
@martinkennelly PTAL

File: pkg/webhook/webhook.go
Comment (martinkennelly) [inline pkg/webhook/webhook.go:639]: should "customizedPatch = append(customizedPatch, patch)" be within the previous if statement closure - before ln 638, so we only add new patches?

**Review by martinkennelly — COMMENTED** [review]
LGTM overall - just one small question and a nit.

File: pkg/webhook/webhook.go
Comment (zshi-redhat) [reply pkg/webhook/webhook.go]: Done

File: pkg/webhook/webhook.go
Comment (zshi-redhat) [reply pkg/webhook/webhook.go]: Yes, it should.
Done

**Review by adrianchiris — COMMENTED** [review]
I still need to give this PR a deeper look.

it would be great if:
1. there is some documentation on this feature in README
2. commit message can be more expressive on the what/how

**Review by martinkennelly — APPROVED** [review]

**Comment by zshi-redhat** [comment]
> I still need to give this PR a deeper look.
> 
> it would be great if:
> 
> 1. there is some documentation on this feature in README
> 2. commit message can be more expressive on the what/how

Done

File: README.md
Comment (adrianchiris) [reply README.md:187]: missing info here on the format of the injection, there is just an example below but i think a more formal definition is needed.

is it in the format as defined by https://tools.ietf.org/html/rfc6902 ?

Comment (adrianchiris) [reply README.md:187]: In addition you should also state that these user defined patches take precedence over built-in patch operations performed by NRI

File: README.md
Comment (adrianchiris) [inline README.md:201]: nit: contain this  label with a value set to `true`

File: pkg/webhook/webhook.go
Comment (adrianchiris) [inline pkg/webhook/webhook.go:641]: can this validation happen on set of `customizedInjections`  instead of every pod creation?

**Review by adrianchiris — CHANGES_REQUESTED** [review]
Thank for proposing this patch Zenghui !

added some comments,

The newly defined methods in webhook.go are fairly isolated, so dare i ask to add unit tests for them ?

File: pkg/webhook/webhook.go
Comment (zshi-redhat) [reply pkg/webhook/webhook.go:707]: customized patch is evaluated first, even there is no built-in patches are required.

File: pkg/webhook/webhook.go
Comment (zshi-redhat) [reply pkg/webhook/webhook.go:66]: could you be specific to the unexpected result? my understanding is if there is a wrong format in customized injection , or incomplete patch or unsupported patch exists in customized Injection, it would cause the createCustomizedPatch fail, the end result would be no customized patch be applied to pod.

File: cmd/webhook/main.go
Comment (zshi-redhat) [reply cmd/webhook/main.go:65]: Yes, kube-system is the default namespace, will update server.yam

File: pkg/webhook/webhook.go
Comment (zshi-redhat) [reply pkg/webhook/webhook.go:641]: `customizedPatch` contains only patch that a pod requests. in the for loop above, there is check on which labels are requested by pod and only add the requested customized patch in `customizedPatch`.
I assume we cannot pre-predict which labels user may set on pod.

File: README.md
Comment (zshi-redhat) [reply README.md:187]: The order of injection may depend on what customized patch are configured by user. For current annotation, it takes precedence then built-in patch, because there is a chance that annotation may contain net-attach-def definition, that would impact the injeciton of resourceName. I'll add the precedence info.

The format is json string, within which, the value structure is an internal data structure defined at: https://github.com/k8snetworkplumbingwg/network-resources-injector/blob/master/pkg/webhook/webhook.go#L43

Comment (adrianchiris) [reply README.md:187]: > The order of injection may depend on what customized patch are configured by user. For current annotation, it takes precedence then built-in patch, because there is a chance that annotation may contain net-attach-def definition, that would impact the injeciton of resourceName. I'll add the precedence info.

Great, Thanks !


> The format is json string, within which, the value structure is an internal data structure defined at:

So The README section should contain this information IMO (the format) as when a user goes through this section he should understand how to write these custom patches which does match RFC6902

File: pkg/webhook/webhook.go
Comment (adrianchiris) [reply pkg/webhook/webhook.go:66]: > could you be specific to the unexpected result?

`SetCustomizedInjection()` first adds new entries then removes stale entires.
As an example for what might go wrong, a user defines a new config replacing all entries, the map is then updated with new configs added.
Then a pod is created, leading to the injections of both new and old configs and only then the "remove stale entries" part in `SetCustomizedInjection()` takes place.

In general you have a data structure that is written and read asynchronously in a non atomic manner by different goroutines, it is generally best practice to add some sort of synchronization mechanism to avoid any inconsistencies when reading the data structure.

File: pkg/webhook/webhook.go
Comment (adrianchiris) [reply pkg/webhook/webhook.go:641]: it is necessarily a subset of `customizedInjections ` isnt it ?
so validating that each patch provided in `customizedInjections` once when setting it will not require to check it per Pod.
it seems to me more efficient.

This does not affect the correctness of the code, and is a suggestion on how to make it a bit more efficient as less operations will be performed when processing a Pod.

File: pkg/webhook/webhook.go
Comment (adrianchiris) [reply pkg/webhook/webhook.go:707]: In the code i see that customized patches are applied only if default network annotation exists or secondary network annotation exists in either the original pod annotation OR as a result of a user-defined labels with a customized injection matching these labels.

Is that the intended behaviour ?
Or should the customized Patch be applied to the pod regardless ?

File: README.md
Comment (zshi-redhat) [reply README.md]: Added RFC6902 reference link.

File: README.md
Comment (zshi-redhat) [reply README.md]: Updated to `User Defined Injections` for this title.

File: cmd/webhook/main.go
Comment (zshi-redhat) [reply cmd/webhook/main.go:65]: Done.

File: pkg/webhook/webhook.go
Comment (zshi-redhat) [reply pkg/webhook/webhook.go]: Added a read/write lock for global structure data update.

File: pkg/webhook/webhook.go
Comment (zshi-redhat) [reply pkg/webhook/webhook.go:705]: This is intended behaviour, at least for now. We want to inject the patchs only for pod requesting resources.

File: README.md
Comment (zshi-redhat) [reply README.md]: Done.

File: pkg/webhook/webhook.go
Comment (zshi-redhat) [reply pkg/webhook/webhook.go]: Moved this down to `SetCustomizedInjection()` so that it only be executed once.

**Comment by zshi-redhat** [comment]
> 
> The newly defined methods in webhook.go are fairly isolated, so dare i ask to add unit tests for them ?

unit tests added in second commit.

File: README.md
Comment (martinkennelly) [inline README.md:185]: nit: additinal -> additional

File: README.md
Comment (martinkennelly) [inline README.md:246]: awesome description - thanks!

File: pkg/webhook/webhook.go
Comment (martinkennelly) [inline pkg/webhook/webhook.go:643]: do you think its worthwhile to cast podValue to lowercase characters before comparison with true?

Comment (zshi-redhat) [reply pkg/webhook/webhook.go]: Yes, it'd be better to use lowercase.
Updated.

File: README.md
Comment (zshi-redhat) [reply README.md]: Good catch!
Done.

**Review by martinkennelly — APPROVED** [review]

**Comment by zshi-redhat** [comment]
@adrianchiris PTAL on the latest changes.

File: README.md
Comment (adrianchiris) [inline README.md:185]: IMO if user defined injections is used, we should use it all the way.
in the content of this section, examples and code. The word "customized" should be replaced with "user defined" so we are consistent throughout.

File: pkg/webhook/webhook.go
Comment (adrianchiris) [inline pkg/webhook/webhook.go:912]: The limitation we impose on the path should be mentioned in the README

File: pkg/webhook/webhook.go
Comment (adrianchiris) [inline pkg/webhook/webhook.go:930]: you are modifying the map at both L#917 and L#926 as `cusPatchs`  is a reference to the original map

both add and delete operations are not atomic to my knowledge so you should either lock/unlock at the critical points or lock at the beginning and defer unlock.

File: pkg/webhook/webhook.go
Comment (adrianchiris) [inline pkg/webhook/webhook.go:932]: no need to do this assignment.

https://play.golang.org/p/pHs2-7u6VXW

**Review by adrianchiris — CHANGES_REQUESTED** [review]

File: pkg/webhook/webhook.go
Comment (zshi-redhat) [reply pkg/webhook/webhook.go:916]: Done

File: pkg/webhook/webhook.go
Comment (zshi-redhat) [reply pkg/webhook/webhook.go]: Done

File: pkg/webhook/webhook.go
Comment (zshi-redhat) [reply pkg/webhook/webhook.go]: Done

File: pkg/webhook/webhook_test.go
Comment (adrianchiris) [inline pkg/webhook/webhook_test.go:359]: Could you add a test that ensures a patch operation fails if the Path is not prefixed with "/metadata/" ?
we should ensure this does not change unintentionally in the future.

File: cmd/webhook/main.go
Comment (adrianchiris) [inline cmd/webhook/main.go:35]: it would be nice if we didnt have the word "customized" in code since we are referring to this as "user-defined"

e.g `customizedInjectionConfigMap` -> `userDefinedInjectionConfigMap` and so on.

not something i would block on, your call if you want to do the extra refactor :)

**Review by adrianchiris — COMMENTED** [review]
Thanks for addressing my comments.

small nit on the unit-tests, the other comment is just a suggestion (optional)

File: cmd/webhook/main.go
Comment (zshi-redhat) [reply cmd/webhook/main.go]: Done, it's easy to update the name than the code logic :-)

File: pkg/webhook/webhook_test.go
Comment (zshi-redhat) [reply pkg/webhook/webhook_test.go:359]: This unit test would fail if other path is used:
https://github.com/k8snetworkplumbingwg/network-resources-injector/pull/59/files#diff-54553ccca2e329f7bf1fdf04d002ced7323d3fa477e7fc6e2a0a9d248cb10ab2R313-R322
do you think it is enough?

Comment (adrianchiris) [reply pkg/webhook/webhook_test.go:359]: Ah, missed this one !

**Review by adrianchiris — APPROVED** [review]
Thanks for addressing my comments !

**Comment by adrianchiris** [comment]
@zshi-redhat mind rebasing this ? once we are green this can be merged IMO

**Comment by zshi-redhat** [comment]
> @zshi-redhat mind rebasing this ? once we are green this can be merged IMO

Rebased

**Pull request merged**
